### ISSUE TITLE
IsEmpty slow, even slower than Count

### ISSUE THREAD

**Issue by nkreipke** [rethinkdb/rethinkdb#4592]
The IsEmpty query is very slow, so that checking if the table is empty using Count and comparing that to 0 is actually more than 3 times faster for large tables.

I confirmed this in Ruby, running each query a hundred times and printing the total time taken:

     def time(count)
        t1 = Time.now
        count.times { yield }
        (Time.now - t1) * 1000 # milliseconds
    end
    
    # Table with 40000 documents
    time(100) { r.db('db').table('table').is_empty.run }
    => 5687.1179999999995
    time(100) { r.db('db').table('table').count.run }
    => 1569.827
    
    # Empty table
    time(100) { r.db('db').table('table').is_empty.run }
    => 226.207
    time(100) { r.db('db').table('table').count.run }
    => 204.003

Shouldn't the time needed to run an IsEmpty query be "fairly" constant, or is there an implementation detail I don't know?

**Comment by danielmewes** [comment]
This is interesting and indeed pretty surprising. Thanks for reporting your findings @nkreipke .

`is_empty` does load exactly one batch worth of data onto the server that the query is run on. If the documents are large or need to be loaded from disk that might be what's making it slow, though 5.6 seconds sound *very* slow.

We should probably use a special smaller batch size for `is_empty`.

@nkreipke To confirm that tuning the batch configuration internally would make `is_empty` faster, could you try the following on your test data set?
```
# Table with 40000 documents
time(100) { r.db('db').table('table').is_empty.run(:max_batch_rows => 1) }
```

**Comment by nkreipke** [comment]
Thanks @danielmewes, that did the trick. The query now takes approx. 2ms on the large table. This should definitely be the default for `is_empty`.

**Comment by mlucy** [comment]
I agree, I think we should just adjust the default batch size for that term.

**Comment by danielmewes** [comment]
Thanks for trying this @nkreipke .

The fix is in code review 3103.

**Comment by danielmewes** [comment]
Merged into `next` https://github.com/rethinkdb/rethinkdb/commit/819d7f3ade3f0ae3fdb3446b93cea96f1e1e6a67 and cherry-picked into `v2.0.x` https://github.com/rethinkdb/rethinkdb/commit/ac9b5c8cb09fcb4162ac951779d28e9641f8c3d3 .

The fix will ship with RethinkDB 2.1.0
